src/edu/washington/escience/myria/operator/SampledDbInsertTemp.java

```diff
@@ -0,0 +1,221 @@
+      accessMethod.tupleBatchInsert(countRelationKey, tb);
+    }
+    return null;
+  }
+
+  /**
+   * Fills reservoir with child tuples.
+   *
+   * @throws DbException
+   *           if TupleBatch fails to get nextReady
+   */
+  private void fillReservoir() throws DbException {
+    Random rand = new Random();
+    for (TupleBatch tb = getChild().nextReady(); tb != null; tb = getChild()
+        .nextReady()) {
+      final List<? extends Column<?>> columns = tb.getDataColumns();
+      for (int i = 0; i < tb.numTuples(); i++) {
+        if (reservoir.numTuples() < streamSampleSize) {
+          // Reservoir size < k. Add this tuple.
+          for (int j = 0; j < tb.numColumns(); j++) {
+            reservoir.put(j, columns.get(j), i);
+          }
+        } else {
+          // Replace probabilistically
+          int replaceIdx = rand.nextInt(tupleCount);
+          if (replaceIdx < reservoir.numTuples()) {
+            for (int j = 0; j < tb.numColumns(); j++) {
+              reservoir.replace(j, replaceIdx, columns.get(j), i);
+            }
+          }
+        }
+        tupleCount++;
+      }
+    }
+    doneSamplingFromChild = true;
+  }
+
+  @Override
+  protected void init(final ImmutableMap<String, Object> execEnvVars)
+      throws DbException {
+    reservoir = new MutableTupleBuffer(getChild().getSchema());
+    /*
+     * retrieve connection information from the environment variables, if not
+     * already set
+     */
+    if (connectionInfo == null && execEnvVars != null) {
+      connectionInfo = (ConnectionInfo) execEnvVars
+          .get(MyriaConstants.EXEC_ENV_VAR_DATABASE_CONN_INFO);
+    }
+
+    if (connectionInfo == null) {
+      throw new DbException(
+          "Unable to instantiate SampledDbInsertTemp: connection information unknown");
+    }
+
+    if (connectionInfo instanceof SQLiteInfo) {
+      /* Set WAL in the beginning. */
+      final File dbFile = new File(
+          ((SQLiteInfo) connectionInfo).getDatabaseFilename());
+      SQLiteConnection conn = new SQLiteConnection(dbFile);
+      try {
+        conn.open(true);
+        conn.exec("PRAGMA journal_mode=WAL;");
+      } catch (SQLiteException e) {
+        e.printStackTrace();
+      }
+      conn.dispose();
+    }
+
+    /* open the database connection */
+    accessMethod = AccessMethod.of(connectionInfo.getDbms(), connectionInfo,
+        false);
+    accessMethod.dropTableIfExists(sampleRelationKey);
+    accessMethod.dropTableIfExists(countRelationKey);
+    // Create the temp tables.
+    accessMethod.createTableIfNotExists(sampleRelationKey, getSchema());
+    accessMethod.createTableIfNotExists(countRelationKey, COUNT_SCHEMA);
+  }
+
+  @Override
+  public void cleanup() {
+    reservoir = null;
+    batches = null;
+
+    try {
+      if (accessMethod != null) {
+        accessMethod.close();
+      }
+    } catch (DbException e) {
+      throw new RuntimeException(e);
```

Comment:
I realize that this is pasted from the other DBInsert operators, but I'm not sure why we need to wrap this in a `RuntimeException`, given that `Operator.close()` can deal with `DbException` being thrown from `cleanup()` (and `cleanup()` explicitly throws `Exception`, so can throw any checked exception). @jingjingwang could you comment?